### ISSUE TITLE
Parse query strings with the router

### ISSUE THREAD

**Issue by MasseGuillaume** [japgolly/scalajs-react#372]
currently, it's not possible to use the router for query params (ex: /foo/bar?q=baz).

https://gitter.im/japgolly/scalajs-react?at=591a1fba83cb5db07334b7b8

**Comment by japgolly** [comment]
FYI the reason this isn't currently supported is because I don't personally need it and it costs me money to work on this lib; it's not because it was hard or I hit a brick wall or anything so if anyone's keen to add this feature, I don't imagine it would be difficult.

**Comment by japgolly** [comment]
Hi @MasseGuillaume, not exactly what you want - in fact it's the opposite but in the new release v1.1.0 there is a new built-in rewrite rule called `removeQuery` which will remove the query string entirely. Maybe that's helpful for you so ping!

**Comment by nafg** [comment]
Is it currently _impossible_ or just tedious? I need this too. I would send a PR if I knew what to do.

**Comment by japgolly** [comment]
@nafg Definitely not impossible, probably not even tedious. If you add some logic to extract the params to a `Map[String, String]` it's going to be quite quick and easy, but to make it effective in real-world usage you'll likely want to add a few more combinators to do things like blacklist/whitelist keys, extract and validate values from the map, etc.

To be clear you'd edit [Dsl.scala](https://github.com/japgolly/scalajs-react/blob/master/extra/src/main/scala/japgolly/scalajs/react/extra/router/Dsl.scala) then, I imagine, add a means to create `RouteB[Map[String, String]]` and use the existing `.map`, `.filter`, etc. to parse & validate keys.

**Comment by japgolly** [comment]
Also read through https://github.com/japgolly/scalajs-react/blob/master/doc/ROUTER.md#routex quickly first

**Comment by kurtome** [comment]
I haven't done extensive testing on with this, but I just had to write a quick solution for myself. This is what I came up with:

```scala
object AdditionalDsl {
  import StaticDsl._

  private val queryCaptureRegex = "(:?(\\?.*)|())#?|$"

  /** Captures the query portion of the URL to a param. */
  def queryToMap: RouteB[Map[String, String]] =
    new RouteB[Map[String, String]](
      queryCaptureRegex,
      1,
      capturedGroups => {
        val capturedQuery = capturedGroups(0).replaceFirst("\\?", "")
        val params = capturedQuery.split("&").filter(_.nonEmpty) map { param =>
          // Note that it is possible for there to be just a key and no value
          val key = param.takeWhile(_ != '=')
          val value = param.drop(key.length + 1)
          key -> value
        }
        Some(params.toMap)
      },
      paramsMap => {
        paramsMap.foldLeft("")((str, param) => {
          if (str.isEmpty) {
            s"?${param._1}=${param._2}"
          } else {
            str + s"&${param._1}=${param._2}"
          }
        })
       }
    )
}
```

With this, example router config like this should work:
```scala
sealed trait MyPages
case object Page1                     extends MyPages
case class  Page2(queryParams: Map[String, String]) extends MyPages
case object Page3                     extends MyPages

val config = RouterConfigDsl[MyPages].buildConfig { dsl =>
  import dsl._
  import AdditionalDsl._
  ( emptyRule
  | staticRoute("page1", Page1) ~> render(???)
  | dynamicRouteCT(("page2" ~ queryToMap).option.caseClass[Page2]) ~> render(???)
  // oops! We forgot Page3!!
  ).notFound(???)
}
```

With that `/page2?k1=v1&k2=v2` should leave you with a `Map("k1" -> "v1", "k2" -> "v2")` in your page case class instance.

**Comment by rleibman** [comment]
I actually tried this, and it worked! The only thing I had to change is removing the .option in your example, but other than that... awesome, thanks!
@japgolly Please add this to the project!

**Comment by japgolly** [comment]
I'd love to add this and I'd be happy to copy-paste just from the comment above but the problem is I'd need some tests too. Would anyone reading this be willing to create a bunch of unit tests and combine it with the above into a PR?

**Comment by rleibman** [comment]
I'll take a look at it in a few days, after I'm back from a conference

**Comment by japgolly** [comment]
Closed via #594
